snapcraft/internal/lxd.py

```diff
@@ -40,30 +72,104 @@ def __init__(self, snap_output, tar_filename, project_options):
         self._project_options = project_options
         self._container_name = 'snapcraft-{}'.format(
             petname.Generate(3, '-'))
+        self._find_lxd()
+        self._container = None
+
+    def _find_lxd(self):
+        try:
+            self._client = pylxd.Client()
+        except:
+            try:
+                os.environ['LXD_DIR'] = '/var/snap/lxd/common/lxd'
+                self._client = pylxd.Client()
+            except:
+                raise EnvironmentError(
+                    'The lxd package is not installed, in order to use '
+                    '`cleanbuild` you must install lxd onto your system. '
+                    'Refer to the "Ubuntu Desktop and Ubuntu Server" section '
+                    'on https://linuxcontainers.org/lxd/getting-started-cli/'
+                    '#ubuntu-desktop-and-ubuntu-server to enable a proper '
+                    'setup.')
 
     def _push_file(self, src, dst):
-        check_call(['lxc', 'file', 'push',
-                    src, '{}/{}'.format(self._container_name, dst)])
+        logger.info('Pushing {} to {}'.format(src, dst))
+        with open(src, 'rb') as source_file:
+            data = source_file.read()
+            self._container.files.put(dst, data)
 
     def _pull_file(self, src, dst):
-        check_call(['lxc', 'file', 'pull',
-                    '{}/{}'.format(self._container_name, src), dst])
+        logger.info('Pulling {} from {}'.format(dst, src))
+        with open(dst, 'wb') as destination_file:
+            data = self._container.files.get(src)
+            destination_file.write(data)
 
     def _container_run(self, cmd):
-        check_call(['lxc', 'exec', self._container_name, '--'] + cmd)
+        if not self._container:
+            raise CalledProcessError(-1, cmd)
+        print('Executing {}'.format(cmd))
+        # API call because self._container.execute doesn't expose IO streams
+        response = self._container.api.exec.post(json={
+            'command': cmd,
+            # Make debconf stop asking for user input
+            'environment': {
+                'DEBIAN_FRONTEND': 'noninteractive',
+            },
+            'wait-for-websocket': True,
+            'interactive': False,
+        })
+        self._fds = response.json()['metadata']['metadata']['fds']
+        operation_id = response.json()['operation'].split('/')[-1]
+        self._parsed = parse.urlparse(
+            self._client.api.operations[operation_id].websocket._api_endpoint)
+        self._manager = WebSocketManager()
+        self._pipe(sys.stdin), self._pipe(sys.stdout), self._pipe(sys.stderr)
+        self._manager.start()
+        while len(self._manager) > 0:
+            sleep(.1)
```

Comment:
Is there a possibility this will wait forever? Perhaps we should implement a timeout.